docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+# Kubelet - Eviction Policy
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+This document presents a specification for how the `kubelet` evicts pods when compute resources are too low.
+
+## Motivation
+
+The node needs a mechanism to preserve stability when available compute resources are low.
+
+## Scope of proposal
+
+This proposal defines a pod eviction policy for reclaiming compute resources.
+
+In the first iteration, it focuses on memory; later iterations are expected to cover
+other resources like disk.  The proposal focuses on a simple default core policy
+intended to cover the broadest class of user workloads.
+
+## Eviction Thresholds
+
+The `kubelet` will support the ability to define an eviction threshold when available
+compute resource drops below an administrator defined value.
+
+The `kubelet` will accept a flag to define the eviction threshold as follows:
+
+* `--eviction-threshold=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe per resource eviction thresholds.
+
+For example, an `--eviction-threshold=memory=256Mi` would mean that if the node has less than
+256Mi of available memory, the node may choose to initiate pod eviction.
+
+The node will only support `memory` first, but other resources such as `cpu` or `disk` may be
+added in the future.
+
+If no eviction threshold is specified for a compute resource, the node may choose a sane
+value (TBD) based on the node capacity.  If the eviction threshold is explicitly set to 0
+for a compute resource, no eviction will occur for that resource.
+
+## Eviction Monitoring Interval
+
+The `kubelet` will support monitoring available compute resource usage at specified
+intervals as a means of determining if node compute resource usage has exhausted
+available compute resource below the eviction threshold.
+
+The `kubelet` will accept a flag to define the eviction monitoring interval as follows:
+
+* `--eviction-monitoring-interval=:` - Interval at which the kubelet monitors available
+compute resource to determine if eviction criteria are met (duration).
+
+The default value is to be determined pending analysis on any additional CPU overhead
+incurred by the `kubelet`.
+
+## Eviction Window
+
+The `kubelet` will evict pods from the node when the eviction threshold has been met
+for a specific window of time.  This allows the node to temporarily consume resources
+that go beyond the specified eviction window without inducing eviction unless that
+consumption was sustained.
+
+The `kubelet` will accept a flag to define the eviction window as follows:
+
+* `--eviction-window=:` - Duration of time where the node is able to exceed the
```

Comment:
I always have a problem with soft limits.  Do you really want to evict in the case of a soft limit for a period of time?  If it's not causing harm then why care.